docs/cheat-sheets/java-xxe.mdx

```diff
@@ -0,0 +1,1839 @@
+---
+slug: java-xxe
+toc_max_heading_level: 4
+---
+
+# XML External entity prevention for Java
+
+This is an XML External Entity (XXE) prevention cheat sheet by r2c. It contains code patterns of potential XXE in an application. The recommendations in this cheat sheet pave a safe road for developers that mitigate the possibility of XXE in their code. By following these recommendations, you can be reasonably sure that their code is free of XXE.
+
+## Mitigation summary
+
+It is not easy to summarize Java XML security. The XML standard allows multiple ways to include external content, including but not limited to external Document Type Definitions (DTD), external Entity References to external data, general entity references, and external parameter references. There are also several ways to include external content in XML Schemas and stylesheets.
+
+The Java API for XML Processing (JAXP) contains multiple interfaces to process XML content:
+
+- The Document Object Model (DOM) interface
+- The Simple API for XML (SAX) interface
+- The Streaming API for XML (StAX) interface
+
+Each of these interfaces has its mitigation measures to disable the use of different sources of external content. The inconsistency in availability and even the effect of these measures across the different APIs, combined with the variety of ways to include external content makes it difficult to provide general mitigation recommendations. Many configurations that are secure for one parser are not secure for another.
+However, the following settings have a consistent effect on a subset of parsers:
+
+- For `SAXBuilder`, `SAXParser`, and `XMLReader` use:
+    ```java
+    setAttribute(XMLConstants.ACCESS_EXTERNAL_DTD, "");
+    ```
+- For `DocumentBuilderFactory`, `TransformerFactory`, `SAXTransformer`, `SchemaFactory`, and `Validator` use:
+    ```java
+    setFeature(XMLConstants.FEATURE_SECURE_PROCESSING, true);
+    ```
+- For `SAXReader` the only secure configuration is:
+    ```java
+    setFeature("http://apache.org/xml/features/<b>disallow-doctype-decl", true);
+
+    ```
+
+The complete Semgrep [set of rules](https://semgrep.dev/r?q=java.lang.security.xxe&lang=Java) that covers vulnerabilities mentioned in this cheat sheet is distributed as part of the rules available in the [Team tier](https://semgrep.dev/pricing). This ruleset includes rules to verify these settings as well as several other combinations that are secure.
```

Comment:
@0xDC0DE reacting on https://github.com/returntocorp/semgrep-docs/pull/804#discussion_r1062595874 - let me know whether this seems better:

```suggestion
The complete Semgrep [set of rules](https://semgrep.dev/r?q=java.lang.security.xxe&lang=Java) that covers vulnerabilities mentioned in this cheat sheet is distributed as part of the rules available in the [Team tier](https://semgrep.dev/pricing). This set of rules verifies the settings mentioned in the list above and also includes combinations of several other security features.
```